crates/chain/src/tx_data_traits.rs

```diff
@@ -31,3 +34,104 @@ impl ForEachTxOut for Transaction {
         }
     }
 }
+
+/// Trait that "anchors" blockchain data in a specific block of height and hash.
+///
+/// This trait is typically associated with blockchain data such as transactions.
+pub trait BlockAnchor:
+    core::fmt::Debug + Clone + Eq + PartialOrd + Ord + core::hash::Hash + Send + Sync + 'static
+{
+    /// Returns the [`BlockId`] that the associated blockchain data is "anchored" in.
+    fn anchor_block(&self) -> BlockId;
+}
+
+impl<A: BlockAnchor> BlockAnchor for &'static A {
+    fn anchor_block(&self) -> BlockId {
+        <A as BlockAnchor>::anchor_block(self)
+    }
+}
+
+impl BlockAnchor for (u32, BlockHash) {
+    fn anchor_block(&self) -> BlockId {
+        (*self).into()
+    }
+}
+
+/// Represents a service that tracks the best chain history.
+/// [TODO] How do we ensure the chain oracle is consistent across a single call?
+/// * We need to somehow lock the data! What if the ChainOracle is remote?
+/// * Get tip method! And check the tip still exists at the end! And every internal call
+///   does not go beyond the initial tip.
+pub trait ChainOracle {
+    /// Error type.
+    type Error: core::fmt::Debug;
+
+    /// Returns the block hash (if any) of the given `height`.
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error>;
+
+    /// Determines whether the block of [`BlockId`] exists in the best chain.
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        Ok(matches!(self.get_block_in_best_chain(block_id.height)?, Some(h) if h == block_id.hash))
+    }
+}
+
+// [TODO] We need stuff for smart pointers. Maybe? How does rust lib do this?
+// Box<dyn ChainOracle>, Arc<dyn ChainOracle> ????? I will figure it out
+impl<C: ChainOracle> ChainOracle for &C {
+    type Error = C::Error;
+
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error> {
+        <C as ChainOracle>::get_block_in_best_chain(self, height)
+    }
+
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        <C as ChainOracle>::is_block_in_best_chain(self, block_id)
+    }
+}
+
+/// Represents changes to a [`TxIndex`] implementation.
+pub trait TxIndexAdditions: Default {
+    /// Append `other` on top of `self`.
+    fn append_additions(&mut self, other: Self);
+}
+
+impl<I: Ord> TxIndexAdditions for BTreeSet<I> {
+    fn append_additions(&mut self, mut other: Self) {
+        self.append(&mut other);
+    }
+}
+
+/// Represents an index of transaction data.
+pub trait TxIndex {
+    /// The resultant "additions" when new transaction data is indexed.
+    type Additions: TxIndexAdditions;
```

Comment:
Remove this trait bound and delete `TxIndexAdditions` and do not provide a default impl for `index_tx`. It's not worth creating this extra trait,.

Reply:
How do you propose we persist changes in the index?

Reply:
How does this trait do this I must be missing something.

Do you mean that `IndexedTxGraph` additions needs to be serializable?
Or maybe that currently we append changesets together when staging them before committing them?

I think perhaps wait until we need this method to actually add it. I can see that you might need to create an `Append` trait which adds an `append` method since that seems to be missing from rust core.

Reply:
How about `AddAssign`?

Reply:
This is my attempt: 90525ec799d669a7ee39ada3be258a06cb6adb49